presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+            add(groupId, block, i, BIGINT.getLong(block, i + 1));
+        }
+    }
+
+    @Override
+    public void ensureCapacity(long size)
+    {
+        // i'm not 100% sure about this--when we have a single group, we pass 0.
+        long actualSize = Math.max(numberOfGroups, size);
+        this.numberOfGroups = actualSize;
+        headPointers.ensureCapacity(actualSize);
+        valueHashList.ensureCapacity(actualSize);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return INSTANCE_SIZE
+                + counts.sizeOf()
+                + groupIds.sizeOf()
+                + nextPointers.sizeOf()
+                + valuePositions.sizeOf()
+                + valueHashList.sizeOf()
+                + buckets.sizeOf()
+                + values.getRetainedSizeInBytes()
+                + valueStore.getEstimatedSize()
+                + headPointers.sizeOf();
+    }
+
+    @Override
+    public void serialize(BlockBuilder out)
+    {
+        if (isCurrentGroupEmpty()) {
+            out.appendNull();
+        }
+        else {
+            BlockBuilder blockBuilder = out.beginBlockEntry();
+
+            iterateGroupNodes(currentGroupId, nodePointer -> {
+                checkArgument(nodePointer != NULL, "should never see null here as we exclude in iterateGroupNodesCall");
+                ValueNode valueNode = bucketNodeFactory.createValueNode(nodePointer);
+                valueNode.writeNodeAsRow(values, blockBuilder);
+            });
+
+            out.closeEntry();
+        }
+    }
+
+    @Override
+    public void addAll(TypedHistogram other)
+    {
+        addAll(currentGroupId, other);
+    }
+
+    @Override
+    public void readAllValues(HistogramValueReader reader)
+    {
+        iterateGroupNodes(currentGroupId, nodePointer -> {
+            checkArgument(nodePointer != NULL, "should never see null here as we exclude in iterateGroupNodesCall");
+            ValueNode valueNode = bucketNodeFactory.createValueNode(nodePointer);
+            reader.read(values, valueNode.getValuePosition(), valueNode.getCount());
+        });
+    }
+
+    @Override
+    public TypedHistogram setGroupId(long groupId)
+    {
+        this.currentGroupId = groupId;
+        return this;
+    }
+
+    @Override
+    public Type getType()
+    {
+        return type;
+    }
+
+    @Override
+    public int getExpectedSize()
+    {
+        return bucketId;
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return isCurrentGroupEmpty();
+    }
+
+    @Override
+    public void add(int position, Block block, long count)
+    {
+        checkState(currentGroupId != -1, "setGroupId() not called yet");
+        add(currentGroupId, block, position, count);
+    }
+
+    private void verifyCapacity()
+    {
+        if (nextNodePointer >= maxFill) {
+            rehash();
+        }
+    }
+
+    private static int computeBucketCount(int expectedSize, float mainFillRate)
```

Comment:
it wasn't clear if the workload of the main hash and the valueStore hash would require different fill ratios. as they are params, i think i just typod "main" instead of "max" 

also ValueStore was once nested, so it took mainMaxFillRatio and valueStoreMaxFillRatio

i've renamed to maxFillRatio

Reply:
done